giskard/rag/question_generators/ookb_questions.py

```diff
@@ -0,0 +1,184 @@
+
+QUESTION_FILTER_OUTPUT_EXAMPLE = """
+
+{
+    "can_be_answered": false,
+    "correctness_reason": "The name of the Boston investment bank where Jessica worked is not directly mentioned in the context."
+}
+
+"""
+
+DUMMY_ANSWER = "This question can not be answered by the context. No sufficient information is provided in the context to answer this question."
+
+
+class OutofKnowledgeBaseGenerator(GenerateFromSingleQuestionMixin, _LLMBasedQuestionGenerator):
+    """
+    Out of Knowledge Base question generator that generates questions from a KnowledgeBase.
+
+    Parameters
+    ----------
+    context_neighbors: int, optional
+        Number of context neighbors to use for question generation.
+    context_similarity_threshold: float, optional
+        Similarity threshold to keep neighboring document during question generation.
+    context_window_length: int, optional
+        Context window length of the llm used in the `llm_client` of the generator.
+    llm_client: LLMClient, optional
+        The LLM client to use for question generation. If not specified, a default openai client will be used.
+    llm_temperature: float, optional
+        The temperature to use in the LLM for question generation. The default is 0.5.
+    """
+
+    _OoKB_question_generation_prompt = QAGenerationPrompt(
+        system_prompt=OoKB_PROMPT,
+        example_input=OoKB_QUESTION_EXAMPLE_INPUT,
+        example_output=OoKB_QUESTION_EXAMPLE_OUTPUT,
+    )
+
+    _OoKB_question_check_prompt = QAGenerationPrompt(
+        system_prompt=QUESTION_CHECK_PROMPT,
+        user_input_template=QUESTION_FILTER_INPUT_TEMPLATE,
+        example_input=QUESTION_FILTER_INPUT_EXAMPLE,
+        example_output=QUESTION_FILTER_OUTPUT_EXAMPLE
+    )
+
+    _question_type = "out of scope"
+
+    # TODO: split the following function into two parts,
+
+    # 1. generate one single question by using OoKB_PROMPT with example input and output
+    # 2. check the question by using QUESTION_FILTER_INPUT_TEMPLATE with example input and output
+    # 3. if the question can not be answered by the context, return the Question Object with the DUMMY_ANSWER
+    # 4. if the question can be answered by the context, back to step 1 and generate a new question and check it again until the question can not be answered by the context.
```

Comment:
??

Reply:
Removed, some thoughts during the development.